python/pyarrow/_io.pyx

```diff
@@ -936,6 +936,44 @@ cdef class HdfsFile(NativeFile):
 # ----------------------------------------------------------------------
 # File and stream readers and writers
 
+class _StreamSinkWrapper(object):
+    """
+    Wrapper for writable stream to ensure implements "tell". This will
+    initialize stream at position 0 and increment after each write.
+    """
+
+    def __init__(self, sink):
+        self._sink = sink
+        self._position = 0
+
+    def close(self):
+        self._sink.close()
+
+    def flush(self):
+        self._sink.flush()
+
+    def tell(self):
+        return self._position
+
+    def write(self, data):
+        data = str(data)
+        self._sink.write(data)
+        self._position += len(data)
+
+    @staticmethod
+    def check(sink):
+        import socket
+        if isinstance(sink, socket.socket):
+            # Use a buffer of 0 to flush after each write, otherwise it is
+            # possible to close socket without flushing
+            sink_file = sink.makefile("wb", 0)
```

Comment:
I was kind of on the fence about automatically wrapping a socket object like this, let me know if you prefer to take this out and maybe just require the application to call `makefile` instead